### ISSUE TITLE
Plan building out support for QCSchema JSON

### ISSUE THREAD

**Issue by langner** [cclib/cclib#643]
We write out CJSON, but we should think about eventually supporting the QCSchema variants as well: https://github.com/MolSSI/QCSchema

Bert wrote a converter from NWChem output: https://github.com/wadejong/NWChemOutputToJson

**Comment by langner** [comment]
Yeah. Honestly I'm a little confused by what it entails as well. But I assume my questions will be answered as soon as we try to validate JSON produced by cclib with their tools (not really sure if they exist yet).

**Comment by dgasmith** [comment]
We are in a bit of a flux state where the schema has been integrated into our quantum chemistry archival project, see [here](https://qcfractal.readthedocs.io/en/latest/) for some very preliminary docs. It might be good to hop on the following [slack](
https://join.slack.com/t/qcdb/shared_invite/enQtNDIzNTQ2OTExODk0LWM3OTgxN2ExYTlkMTlkZjA0OTExZDlmNGRlY2M4NWJlNDlkZGQyYWUxOTJmMzc3M2VlYzZjMjgxMDRkYzFmOTE) so that I can get you and a few others together to look into a couple items.

**Comment by langner** [comment]
Thanks for reaching out, Daniel. I have noticed there are a number of related projects now. I am following a few, but am wondering about the scope of the various parts. WIll jump on the slack you linked to.

**Comment by akshansh2000** [comment]
@langner @berquist 
Is this issue up for grabs for GSoC '20?

**Comment by langner** [comment]
> @langner @berquist
> Is this issue up for grabs for GSoC '20?

Yes, very much so!

(it's also on the ideas page: http://wiki.openchemistry.org/GSoC_Ideas_2020#Project:_Support_for_QCSchema_JSON_output)

**Comment by akshansh2000** [comment]
> Yes, very much so!

Ahh, that's great.
So the basic idea is to convert the current outputs (e.g. molcas output) to the QCSchema output, right? I saw an example of the same for a water molecule [here](https://molssi-qc-schema.readthedocs.io/en/latest/examples.html).

Could you provide some pointers as to where to start?

**Comment by ghutchis** [comment]
I would think both a writer and a reader would be desired, since some software is making QCSchema output.

I'm sure @langner and @berquist will chime in, but check out the cjson code in `cclib/io`

**Comment by berquist** [comment]
> I would think both a writer and a reader would be desired, since some software is making QCSchema output.

Yes, we need both a reader and writer. Two approaches currently come to mind:
1. Implement the reader and writer in the style of https://github.com/cclib/cclib/blob/25eb8104f70b658efe3c2c7dbf31049c1122c94a/cclib/io/cjsonreader.py and https://github.com/cclib/cclib/blob/25eb8104f70b658efe3c2c7dbf31049c1122c94a/cclib/io/cjsonwriter.py. The correspondence between attributes and Chemical JSON keys is held in https://github.com/cclib/cclib/blob/25eb8104f70b658efe3c2c7dbf31049c1122c94a/cclib/parser/data.py#L99, and the conversion between our `ccData` format and Chemical JSON is for writing is in https://github.com/cclib/cclib/blob/25eb8104f70b658efe3c2c7dbf31049c1122c94a/cclib/io/cjsonwriter.py#L38. Reading is a bit easier. See https://github.com/cclib/cclib/pull/798 for some stubs of what this might look like. (It used to be called QCJSON.)
2. Rework how we store data so that rather than have the members on a `ccData` instance be simple data structures, we have a class for each general attribute type than can become arbitrarily complex in order to deal with the diversity of program capabilities. This is the essence of https://github.com/cclib/cclib/issues/419. In this approach, rather than have the monolithic `QCSchemaWriter` be coupled to the implementation of every attribute, each attribute knows how to write its contribution to a QCSchema file, a Chemical JSON file, a Molden file, and so on. This would make adding parts of the `QCSchemaWriter` easier, but you'd also be doing the core of the restructure we've been thinking about for cclib 2.0.

In either case,
- we'd need to work on what the correct mapping between `ccData` and QCSchema attributes is
- we should have some documentation for what is missing from our QCSchema implementation for attributes that we can't capture yet, such as [fragments](https://github.com/cclib/cclib/issues/131)
